### PR TITLE
listener support

### PR DESCRIPTION
**Pull request by Jammyson**
listener support

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -55,6 +55,15 @@
             <artifactId>spring-boot-starter-test</artifactId>
             <scope>test</scope>
         </dependency>
+        <dependency>
+            <groupId>com.github.taccisum</groupId>
+            <artifactId>shiro-starter</artifactId>
```
Comment (taccisum) [inline pom.xml]: 不能引入这个组件哈

File: pom.xml [hunk 2]
```diff
@@ -55,6 +55,15 @@
...
+            <version>2.2.0</version>
+        </dependency>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
+            <artifactId>spring-boot-configuration-processor</artifactId>
```
Comment (taccisum) [inline pom.xml]: `<scope>optional</scope>`

File: src/main/java/com/deepexi/support/amqp/listener/model/MQListenerProperties.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * @author Y.H.Zhou
+ * @since 2020/2/3.
+ * <p></p>
+ */
+@ConfigurationProperties("mq-listener-support")
+@Data
+public class MQListenerProperties {
+    private boolean fake;
```
Comment (taccisum) [inline src/main/java/com/deepexi/support/amqp/listener/model/MQListenerProperties.java]: 这个配置在这里干嘛用？

File: src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+    @Autowired
+    private MessageCallbackHandler callbackHandler;
+
+    @RabbitHandler(isDefault = true)
+    public void listen(Object message, @Headers Map headers) {
+        log.warn("handler not found. message[content: {}, headers: {}] will be ignored.", message, headers);
+    }
+
+    protected void withAuthAndRecord(Object data, Map headers, Action action) {
```
Comment (taccisum) [inline src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java]: 不要照搬项目中的代码，项目中的代码是有局限的，提取成组件后要更加考虑通用性。

举个例子，你怎么知道使用你这个组件的用户项目中一定就有认证机制并且一定是按照我们这种方式去认证的呢？

对于这种不确定的行为，正确的做法应该是提供一个钩子函数（模板方法模式），让用户自己去决定要如何执行相关的逻辑。

File: src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+import java.util.Map;
+
+/**
+ * @author Y.H.Zhou
+ * @since 2020/2/3.
+ * <p></p>
+ */
+@Slf4j
+public abstract class AbstractListener {
+    @Autowired
```
Comment (taccisum) [inline src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java]: 别在抽象类里用 autowired

File: src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    private MessageHandler messageHandler;
+    @Autowired
+    private MessageHelper messageHelper;
```
Comment (taccisum) [inline src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java]: message helper 这个概念不太清晰，helper 一般是代表静态工具类

File: src/main/java/com/deepexi/support/amqp/listener/handler/MessageHandler.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+public class MessageHandler {
+    public void preHandle(Message message) {
+
+    }
+
+    public void postHandle(Message message) {
+
+    }
+
+    public void defaultListenerHandle(Message message) {
```
Comment (taccisum) [inline src/main/java/com/deepexi/support/amqp/listener/handler/MessageHandler.java]: 这个方法不是一个处理动作吗，为啥命名成名词

Comment (taccisum) [reply src/main/java/com/deepexi/support/amqp/listener/handler/MessageHandler.java]: 名词方法一般是为了简化写法，其潜在意思是 getXXX，比如 xxx.data()，潜台词是 xxx.getData()

File: src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java (hunk 6 above)
Comment (taccisum) [reply src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java]: 你要考虑使用者在扩展的时候的情况，如果我是用户，我在使用的时候你让我定义一个 `MessageHelper`，我估计得一脸懵圈

File: src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java (hunk 5 above)
Comment (taccisum) [reply src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java]: 继承这个类的类实例，是不是通过 spring 创建出来都还不一定呢

File: src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        String messageId = messageHelper.getMessageId(headers);
+        if (messageHelper.isConsumed(messageId)) {
+            log.warn("message[message id: {}] has been consumed. it will be ignored.", messageId);
+            return;
+        }
+
+        messageHandler.defaultListenerHandle(messageHelper.messageBuilder(headers, message));
+    }
+
+    public final void consume(Map headers, Object data, Action action) {
```
Comment (taccisum) [inline src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java]: 这个方法的命名还是再考虑一下
另外应该是 protected 级别吧？

File: src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+        String messageId = messageHelper.getMessageId(headers);
+        if (messageHelper.isConsumed(messageId)) {
+            log.warn("message[message id: {}] has been consumed. it will be ignored.", messageId);
+            return;
+        }
+
+        Message message = messageHelper.messageBuilder(headers, data);
```
Comment (taccisum) [inline src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java]: `messageHelper .messageBuilder()` 返回的却是一个 `Message`，不觉得很奇怪吗

File: src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java (hunk 8 above)
Comment (taccisum) [reply src/main/java/com/deepexi/support/amqp/listener/AbstractListener.java]: 另外，提供父类模板方法的做法在单个项目内还算常见，但在通用组件中使用起来其实不是特别友好，可以考虑一下是否有别的更好的方式（比如切面、或者调研一下 spring amqp 本身有没有针对 @RabbitHandler 的扩展点）

**Comment by JackyHweng** [comment]
# [Codecov](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=h1) Report
> Merging [#1](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=desc) into [master](https://codecov.io/gh/deepexi/deepexi-amqp-support/commit/63bf5d0a23d53855288660c5100cb271576e6160&el=desc) will **decrease** coverage by `31.64%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=tree)

```diff
@@              Coverage Diff              @@
##             master       #1       +/-   ##
=============================================
- Coverage     60.43%   28.79%   -31.65%     
  Complexity       22       22               
=============================================
  Files             5       14        +9     
  Lines            91      191      +100     
  Branches         13       18        +5     
=============================================
  Hits             55       55               
- Misses           34      134      +100     
  Partials          2        2               
```


| [Impacted Files](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=tree) | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| [.../support/amqp/listener/aspectJ/ListenerAspect.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvYXNwZWN0Si9MaXN0ZW5lckFzcGVjdC5qYXZh) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...rt/amqp/listener/handler/SimpleMessageHandler.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvaGFuZGxlci9TaW1wbGVNZXNzYWdlSGFuZGxlci5qYXZh) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...port/amqp/listener/handler/RabbitErrorHandler.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvaGFuZGxlci9SYWJiaXRFcnJvckhhbmRsZXIuamF2YQ==) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [.../support/amqp/listener/handler/MessageHandler.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvaGFuZGxlci9NZXNzYWdlSGFuZGxlci5qYXZh) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...epexi/support/amqp/listener/AmqpSupportConfig.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvQW1xcFN1cHBvcnRDb25maWcuamF2YQ==) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...eepexi/support/amqp/listener/AbstractListener.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvQWJzdHJhY3RMaXN0ZW5lci5qYXZh) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...mqp/listener/handler/ListenerExtensionHandler.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvaGFuZGxlci9MaXN0ZW5lckV4dGVuc2lvbkhhbmRsZXIuamF2YQ==) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...m/deepexi/support/amqp/listener/MessageHelper.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvTWVzc2FnZUhlbHBlci5qYXZh) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...ava/com/deepexi/support/amqp/listener/Message.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvTWVzc2FnZS5qYXZh) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=footer). Last update [63bf5d0...e8f860a](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by JackyHweng** [comment]
# [Codecov](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=h1) Report
> Merging [#1](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=desc) into [master](https://codecov.io/gh/deepexi/deepexi-amqp-support/commit/63bf5d0a23d53855288660c5100cb271576e6160&el=desc) will **decrease** coverage by `13.82%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=tree)

```diff
@@              Coverage Diff              @@
##             master       #1       +/-   ##
=============================================
- Coverage     60.43%   46.61%   -13.83%     
  Complexity       22       22               
=============================================
  Files             5        8        +3     
  Lines            91      118       +27     
  Branches         13       14        +1     
=============================================
  Hits             55       55               
- Misses           34       61       +27     
  Partials          2        2               
```


| [Impacted Files](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=tree) | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| [.../deepexi/support/amqp/listener/MessageHandler.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvTWVzc2FnZUhhbmRsZXIuamF2YQ==) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...ort/amqp/listener/SimpleMessageHandlerFactory.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvU2ltcGxlTWVzc2FnZUhhbmRsZXJGYWN0b3J5LmphdmE=) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...amqp/listener/DelegatingInvocableHandleMethod.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvRGVsZWdhdGluZ0ludm9jYWJsZUhhbmRsZU1ldGhvZC5qYXZh) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=footer). Last update [63bf5d0...78f9165](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by JackyHweng** [comment]
# [Codecov](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=h1) Report
> Merging [#1](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=desc) into [master](https://codecov.io/gh/deepexi/deepexi-amqp-support/commit/63bf5d0a23d53855288660c5100cb271576e6160&el=desc) will **decrease** coverage by `26.27%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=tree)

```diff
@@              Coverage Diff              @@
##             master       #1       +/-   ##
=============================================
- Coverage     60.43%   34.16%   -26.28%     
  Complexity       22       22               
=============================================
  Files             5       14        +9     
  Lines            91      161       +70     
  Branches         13       17        +4     
=============================================
  Hits             55       55               
- Misses           34      104       +70     
  Partials          2        2               
```


| [Impacted Files](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=tree) | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| [...mqp/listener/InvocableHandlerMethodDecoration.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvSW52b2NhYmxlSGFuZGxlck1ldGhvZERlY29yYXRpb24uamF2YQ==) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...qp/autoconfigure/AmqpSupportAutoConfiguration.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvYXV0b2NvbmZpZ3VyZS9BbXFwU3VwcG9ydEF1dG9Db25maWd1cmF0aW9uLmphdmE=) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...ort/amqp/listener/AbstractIdempotentValidator.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvQWJzdHJhY3RJZGVtcG90ZW50VmFsaWRhdG9yLmphdmE=) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...amqp/listener/InvocableHandlerMethodDecorator.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvSW52b2NhYmxlSGFuZGxlck1ldGhvZERlY29yYXRvci5qYXZh) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...deepexi/support/amqp/listener/MessageRecorder.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvTWVzc2FnZVJlY29yZGVyLmphdmE=) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...exi/support/amqp/listener/IdempotentValidator.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvSWRlbXBvdGVudFZhbGlkYXRvci5qYXZh) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...amqp/listener/SimpleMemoryIdempotentValidator.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvU2ltcGxlTWVtb3J5SWRlbXBvdGVudFZhbGlkYXRvci5qYXZh) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...m/deepexi/support/amqp/listener/Authenticator.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvQXV0aGVudGljYXRvci5qYXZh) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...listener/DecoratedMessageHandlerMethodFactory.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvRGVjb3JhdGVkTWVzc2FnZUhhbmRsZXJNZXRob2RGYWN0b3J5LmphdmE=) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=footer). Last update [63bf5d0...f2fad18](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

**Comment by JackyHweng** [comment]
# [Codecov](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=h1) Report
> Merging [#1](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=desc) into [master](https://codecov.io/gh/deepexi/deepexi-amqp-support/commit/63bf5d0a23d53855288660c5100cb271576e6160&el=desc) will **decrease** coverage by `7.02%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=tree)

```diff
@@             Coverage Diff              @@
##             master       #1      +/-   ##
============================================
- Coverage     60.43%   53.41%   -7.03%     
- Complexity       22       35      +13     
============================================
  Files             5       14       +9     
  Lines            91      161      +70     
  Branches         13       17       +4     
============================================
+ Hits             55       86      +31     
- Misses           34       72      +38     
- Partials          2        3       +1     
```


| [Impacted Files](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=tree) | Coverage Δ | Complexity Δ | |
|---|---|---|---|
| [...ner/decorated/InvocableHandlerMethodDecorator.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvZGVjb3JhdGVkL0ludm9jYWJsZUhhbmRsZXJNZXRob2REZWNvcmF0b3IuamF2YQ==) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...t/amqp/listener/decorator/IdempotentValidator.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvZGVjb3JhdG9yL0lkZW1wb3RlbnRWYWxpZGF0b3IuamF2YQ==) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...istener/decorator/AbstractIdempotentValidator.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvZGVjb3JhdG9yL0Fic3RyYWN0SWRlbXBvdGVudFZhbGlkYXRvci5qYXZh) | `83.33% <0.00%> (ø)` | `4.00% <0.00%> (?%)` | |
| [...support/amqp/listener/decorator/Authenticator.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvZGVjb3JhdG9yL0F1dGhlbnRpY2F0b3IuamF2YQ==) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...ner/decorator/SimpleMemoryIdempotentValidator.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvZGVjb3JhdG9yL1NpbXBsZU1lbW9yeUlkZW1wb3RlbnRWYWxpZGF0b3IuamF2YQ==) | `100.00% <0.00%> (ø)` | `3.00% <0.00%> (?%)` | |
| [...er/decorated/InvocableHandlerMethodDecoration.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvZGVjb3JhdGVkL0ludm9jYWJsZUhhbmRsZXJNZXRob2REZWNvcmF0aW9uLmphdmE=) | `100.00% <0.00%> (ø)` | `2.00% <0.00%> (?%)` | |
| [...pport/amqp/listener/decorator/MessageRecorder.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvZGVjb3JhdG9yL01lc3NhZ2VSZWNvcmRlci5qYXZh) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |
| [...ecorated/DecoratedMessageHandlerMethodFactory.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvbGlzdGVuZXIvZGVjb3JhdGVkL0RlY29yYXRlZE1lc3NhZ2VIYW5kbGVyTWV0aG9kRmFjdG9yeS5qYXZh) | `100.00% <0.00%> (ø)` | `4.00% <0.00%> (?%)` | |
| [...qp/autoconfigure/AmqpSupportAutoConfiguration.java](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1/diff?src=pr&el=tree#diff-c3JjL21haW4vamF2YS9jb20vZGVlcGV4aS9zdXBwb3J0L2FtcXAvYXV0b2NvbmZpZ3VyZS9BbXFwU3VwcG9ydEF1dG9Db25maWd1cmF0aW9uLmphdmE=) | `0.00% <0.00%> (ø)` | `0.00% <0.00%> (?%)` | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=footer). Last update [63bf5d0...b79e6e4](https://codecov.io/gh/deepexi/deepexi-amqp-support/pull/1?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: src/main/java/com/deepexi/support/amqp/listener/decorator/SimpleMemoryIdempotentValidator.java [hunk 10]
```diff
@@ -0,0 +1,29 @@
+package com.deepexi.support.amqp.listener.decorator;
+
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * @author taccisum - liaojinfeng@deepexi.com
+ * @since 2020-02-21
+ */
+public class SimpleMemoryIdempotentValidator extends AbstractIdempotentValidator {
```
Comment (taccisum) [inline src/main/java/com/deepexi/support/amqp/listener/decorator/SimpleMemoryIdempotentValidator.java:10]: 补充一下这个类的单元测试

File: src/main/java/com/deepexi/support/amqp/listener/decorator/AbstractIdempotentValidator.java [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.amqp.support.AmqpHeaders;
+import org.springframework.messaging.Message;
+
+/**
+ * @author taccisum - liaojinfeng@deepexi.com
+ * @since 2020-02-21
+ */
+@Slf4j
+public abstract class AbstractIdempotentValidator implements IdempotentValidator {
```
Comment (taccisum) [inline src/main/java/com/deepexi/support/amqp/listener/decorator/AbstractIdempotentValidator.java:12]: 补充一下这个类的单元测试

**Pull request merged**